### PR TITLE
Added possible FSM implementation

### PR DESCRIPTION
**Pull request by tituschow**
I'm putting this FSM implementation up for review. I haven't decided whether to make it a strict Moore machine, but it would be easy to do so. Every time an event is processed and a transition occurs, the next state's output function is called.

My goals for this FSM library are to make understanding and adding to the transition tables easier. I decided to accomplish this by declaring my transition tables as functions and hiding that away using macros, so the State objects could be treated as opaque structs. This implementation does not support composite states.

Let me know what you think.

### CODE REVIEW

File: event_queue.h [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+#define EVENT_QUEUE_SIZE 20
+
+typedef struct Event {
+  uint16_t id;
+  uint16_t data;
+} Event;
+
+// Initializes the event queue.
+void event_queue_init(void);
```
Comment (ckitagawa) [inline event_queue.h]: Where is the implementation of this?

File: event_queue.h [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+// Raises an event in the global event queue.
+bool event_raise(Event *e);
```
Comment (ckitagawa) [inline event_queue.h]: Nitpicky but raise_event() makes more sense to me....

File: fsm.h [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+// Usage - Forward declare all states with FSM_DECLARE_STATE:
+//
+// FSM_DECLARE_STATE(state_a);
+// FSM_DECLARE_STATE(state_b);
+//
+// Then define the transition table for each state:
+//
+// FSM_STATE_TRANSITION(state_a) {
+//   FSM_ADD_TRANSITION(0, state_b);
+// }
```
Comment (ckitagawa) [inline fsm.h]: I really like this interface for configuring states

File: fsm.h [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+  const char *name;
+  StateOutput output;
+  StateTransition table;
+} State;
+
+typedef struct FSM {
+  const char *name;
+  State *last_state;
+  State *current_state;
+} FSM;
```
Comment (ckitagawa) [inline fsm.h]: Named twice?

File: fsm.h [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+struct FSM;
+typedef void (*StateOutput)(struct FSM *fsm, const Event *e);
+typedef void (*StateTransition)(struct FSM *fsm, const Event *e, bool *transitioned);
+
+typedef struct State {
+  const char *name;
+  StateOutput output;
+  StateTransition table;
+} State;
```
Comment (ckitagawa) [inline fsm.h]: So we are allowing typedef struct now? Cool. Why do you name the struct twice? Once is sufficient unless I am really missing something in our styleguide.

File: fsm_internal.h [hunk 6]
```diff
@@ -0,0 +1,36 @@
+#pragma once
```
Comment (ckitagawa) [inline fsm_internal.h]: These macros are kind of gross. I understand why they are needed but I kind of wish we have template functions or something. These should all be documented more thoroughly so first-years can follow.

File: main.c [hunk 7]
```diff
@@ -0,0 +1,61 @@
+#include <stdio.h>
```
Comment (ckitagawa) [inline main.c]: You should just make a unit test for this...

**Review by ckitagawa — COMMENTED** [review]
Overall I really like this as a design, as for the Moore vs Mealy I don't have a strong preference. As it is I feel like it is very easy to understand and quite usable. The issue is if the missing functionality from the old FSM will hinder design.

File: event_queue.h (hunk 1 above)
Comment (tituschow) [reply event_queue.h]: This is just an interface for now. I'll implement the event queue later.

File: event_queue.h (hunk 2 above)
Comment (tituschow) [reply event_queue.h]: As part of our coding standard, we prefix all functions with their module name. It just helps organization better.

File: fsm.h (hunk 5 above)
Comment (tituschow) [reply fsm.h]: It allows you to refer to it as a struct, such as when I needed to forward declare it.

File: fsm.h (hunk 4 above)
Comment (tituschow) [inline fsm.h]: See comment on struct

File: main.c (hunk 7 above)
Comment (tituschow) [reply main.c]: Yeah, this was just an example program so people could see how it's used.

File: fsm_internal.h (hunk 6 above)
Comment (tituschow) [reply fsm_internal.h]: Haha, fair. They aren't that bad once you remove all the backslashes imo :P
I'm a little torn on the level of documentation needed for the macros. I've already explained the general idea of how they work, and I don't want to end up duplicating the code as a comment.

File: event_queue.h (hunk 2 above)
Comment (ckitagawa) [reply event_queue.h]: Makes sense since we don't have cpp namespaces

**Review by ckitagawa — APPROVED** [review]
LGTM

**Pull request merged**